test/automation/src/positron/positronNotebooks.ts

```diff
@@ -75,40 +76,27 @@ export class PositronNotebooks {
 	}
 
 	async addCodeToFirstCell(code: string) {
-		await this.code.driver.getLocator(CELL_LINE).first().click();
+		await this.code.driver.page.locator(CELL_LINE).first().click();
 		await this.notebook.waitForTypeInEditor(code);
 		await this.notebook.waitForActiveCellEditorContents(code);
 	}
 
 	async executeCodeInCell() {
 		await this.quickaccess.runCommand(EXECUTE_CELL_COMMAND);
+		await expect(this.code.driver.page.locator(EXECUTE_CELL_SPINNER)).not.toBeVisible({ timeout: 30000 });
 	}
 
-	async getPythonCellOutput(): Promise<string> {
-		// basic CSS selection doesn't support frames (or nested frames)
-		const notebookFrame = this.code.driver.getFrame(OUTER_FRAME).frameLocator(INNER_FRAME);
-		const outputLocator = notebookFrame.locator(PYTHON_OUTPUT);
-		const outputText = await outputLocator.textContent();
-		return outputText!;
-	}
-
-	async getRCellOutput(): Promise<string> {
-		// basic CSS selection doesn't support frames (or nested frames)
-		const notebookFrame = this.code.driver.getFrame(OUTER_FRAME).frameLocator(INNER_FRAME);
-		const outputLocator = notebookFrame.locator(R_OUTPUT).nth(0);
-		const outputText = await outputLocator.textContent();
-		return outputText!;
+	async assertCellOutput(text: string): Promise<void> {
+		await expect(this.frameLocator.getByText(text)).toBeVisible();
 	}
 
 	async closeNotebookWithoutSaving() {
 		await this.quickaccess.runCommand(REVERT_AND_CLOSE);
 	}
 
-	async getMarkdownText(tag: string): Promise<string> {
-		// basic CSS selection doesn't support frames (or nested frames)
-		const frame = this.code.driver.getFrame(OUTER_FRAME).frameLocator(INNER_FRAME);
-		const element = frame.locator(`${MARKDOWN_TEXT} ${tag}`);
-		const text = await element.textContent();
```

Comment:
just an example, but this is what was problematic. the text is being retrieved immediately, which means it might grab the content before it’s fully rendered or updated, causing the assertion to happen before the text is actually correct/ready.